### ISSUE TITLE
Defining an infix operator `!` is recognised as a bang pattern

### ISSUE THREAD

**Issue by Kritzefitz** [ndmitchell/hlint#808]
I tried to define an operator called `(!)` like this:

```Haskell
(!) :: a -> b -> ()
_ ! _ = ()
```
But hlint gives me a parse error, while GHC accepts the codes just fine:
```
test.hs:4:1: Error: Parse error
Found:
    (!) :: a -> b -> ()
  > _ ! _ = ()
```

I think hlint (haskell-src-exts?) thinks `! _` is a bang pattern and thus expects the first `_` to be a valid function name, as demonstrated by this code:

```Haskell
(!) :: a -> () -> ()
a ! () = ()
```
which gives the following hint:
```
test.hs:4:3: Warning: Redundant bang pattern
Found:
  !()
Perhaps:
  ()
```

**Comment by ndmitchell** [comment]
Thanks for the report. We're working on the conversion to the GHC API so hopefully this bug will be fixed by that.

**Comment by ndmitchell** [comment]
This is still an error in the new world, since we turn on bang patterns, and they break this code.

**Comment by ndmitchell** [comment]
I believe GHC HEAD changes the way such operators are parsed, so this might go away with the next GHC.

**Comment by shayne-fletcher** [comment]
So the rules for this one are, when the occurrence is prefix it's a strictness annotation or bang pattern e.g. ` f !x = rhs`, `data T = MkT !a`. Any other usage is an ordinary operator or type operator e.g. ` xs ! 3`, `(! x)`, `Int ! Bool`. With ghc-8.10, the above program in the presence of `BangPatterns` gives a parse error, with GHC HEAD, the issue resolved and a successful parse. So yes, the problem will go away with the next GHC.
